File: python/xgrammar/structural_tag.py
"""Defines all structural tag formats."""

import json
from typing import Any, Dict, List, Literal, Type, Union

try:
    # Python 3.9+
    from typing import Annotated
except ImportError:
    # Python 3.8
    from typing_extensions import Annotated

from pydantic import BaseModel, Field

# ---------- Basic Formats ----------


class ConstStringFormat(BaseModel):
    """A format that matches a constant string."""

    type: Literal["const_string"] = "const_string"
    """The type of the format."""
    value: str
    """The constant string."""


class JSONSchemaFormat(BaseModel):
    """A format that matches a JSON schema."""

    type: Literal["json_schema"] = "json_schema"
    """The type of the format."""
    json_schema: Union[bool, Dict[str, Any]]
    """The JSON schema."""
    style: Literal["json", "qwen_xml", "minimax_xml", "deepseek_xml", "glm_xml"] = "json"
    """How to parse the content. Valid values: \"json\" (standard JSON), \"qwen_xml\" (Qwen XML:
    <parameter=key>value</parameter>), \"minimax_xml\" (MiniMax XML: <parameter name=\"key\">value</parameter>),
    \"deepseek_xml\" (DeepSeek XML(DeepSeek-v3.2): <{dsml_token}parameter name=\"key\" string=\"true|false\">value</{dsml_token}parameter>),
    \"glm_xml\" (GLM XML: <arg_key>key</arg_key><arg_value>value</arg_value>)."""


class QwenXMLParameterFormat(BaseModel):
    """A format that matches Qwen XML function calls.

    Examples
    --------
    .. code-block:: python

        structural_tag = QwenXMLParameterFormat(
            json_schema={
                "type": "qwen_xml_parameter",
                "json_schema": {
                    "type": "object",
                    "properties": {"name": {"type": "string"}, "age": {"type": "integer"}},
                    "required": ["name", "age"],
                },
            }
        )

    The above structural tag can accept the following outputs::

        <parameter=name>Bob</parameter><parameter=age>100</parameter>
        <parameter=name>"Bob&lt;"</parameter><parameter=age>100</parameter>

    """

    type: Literal["qwen_xml_parameter"] = "qwen_xml_parameter"
    """The type of the format."""

    json_schema: Union[bool, Dict[str, Any]]
    """The JSON schema for the parameters of the function calling."""


class AnyTextFormat(BaseModel):
    """A format that matches any text."""

    type: Literal["any_text"] = "any_text"
    """The type of the format."""

    excludes: List[str] = []
    """List of strings that should not appear in the matched text."""


class TokenFormat(BaseModel):
    """A format that matches a single token by ID or string representation."""

    type: Literal["token"] = "token"
    """The type of the format."""

    token: Union[int, str]
    """The token ID (int) or token string (str)."""


class ExcludeTokenFormat(BaseModel):
    """A format that matches a single token, excluding those in the given set."""

    type: Literal["exclude_token"] = "exclude_token"
    """The type of the format."""

    exclude_tokens: List[Union[int, str]] = []
    """List of token IDs or strings to exclude."""


class AnyTokensFormat(BaseModel):
    """A format that matches zero or more tokens, excluding those in the given set."""

    type: Literal["any_tokens"] = "any_tokens"
    """The type of the format."""

    exclude_tokens: List[Union[int, str]] = []
    """List of token IDs or strings to exclude."""


class GrammarFormat(BaseModel):
    """A format that matches an ebnf grammar."""

    type: Literal["grammar"] = "grammar"
    """The type of the format."""

    grammar: str
    """The ebnf grammar."""


class RegexFormat(BaseModel):
    """A format that matches a regex pattern."""

    type: Literal["regex"] = "regex"
    """The type of the format."""

    pattern: str
    """The regex pattern."""


# ---------- Combinatorial Formats ----------


class SequenceFormat(BaseModel):
    """A format that matches a sequence of formats."""

    type: Literal["sequence"] = "sequence"
    """The type of the format."""
    elements: List["Format"]
    """The elements of the sequence."""


class OrFormat(BaseModel):
    """A format that matches one of the formats."""

    type: Literal["or"] = "or"
    """The type of the format."""
    elements: List["Format"]
    """The elements of the or."""


class TagFormat(BaseModel):
    """A format that matches a tag: ``begin content end``.

    The ``end`` field can be a single string or a list of possible end strings.
    When multiple end strings are provided, any of them will be accepted as a valid
    ending for the tag.

    Examples
    --------

    Single end string:

    .. code-block:: python

        TagFormat(begin="<response>", content=..., end="</response>")

    Multiple end strings:

    .. code-block:: python

        TagFormat(begin="<response>", content=..., end=["</response>", "</answer>"])

    """

    type: Literal["tag"] = "tag"
    """The type of the format."""
    begin: Union[str, TokenFormat]
    """The begin tag. Can be a string or a TokenFormat."""
    content: "Format"
    """The content of the tag. It can be any of the formats."""
    end: Union[str, List[str], TokenFormat]
    """The end tag(s). Can be a string, list of strings, or a TokenFormat."""


class TriggeredTagsFormat(BaseModel):
    """A format that matches triggered tags. It can allow any output until a trigger is
    encountered, then dispatch to the corresponding tag; when the end tag is encountered, the
    grammar will allow any following output, until the next trigger is encountered.

    Each tag should be matched by exactly one trigger. "matching" means the trigger should be a
    prefix of the begin tag.

    Tags must use **string** ``begin`` fields, not ``TokenFormat``.
    For token-level dispatch, use ``TokenTriggeredTagsFormat`` instead.

    Examples
    --------

    .. code-block:: python

        structural_tag = TriggeredTagsFormat(
            triggers=["<function="],
            tags=[
                TagFormat(
                    begin="<function=func1>",
                    content=JSONSchemaFormat(json_schema=...),
                    end="</function>",
                ),
                TagFormat(
                    begin="<function=func2>",
                    content=JSONSchemaFormat(json_schema=...),
                    end="</function>",
                ),
            ],
            at_least_one=False,
            stop_after_first=False,
        )

    The above structural tag can accept the following outputs::

        <function=func1>{"name": "John", "age": 30}</function>
        <function=func2>{"name": "Jane", "age": 25}</function>
        any_text<function=func1>{"name": "John", "age": 30}</function>any_text1<function=func2>{"name": "Jane", "age": 25}</function>any_text2

    """

    type: Literal["triggered_tags"] = "triggered_tags"
    """The type of the format."""
    triggers: List[str]
    """The triggers of the triggered tags."""
    tags: List[TagFormat]
    """The tags of the triggered tags."""
    at_least_one: bool = False
    """Whether at least one of the tags must be generated."""
    stop_after_first: bool = False
    """Whether to stop after the first tag is generated."""
    excludes: List[str] = []
    """List of strings that should not appear in the matched text."""


class TokenTriggeredTagsFormat(BaseModel):
    """A format that dispatches to tags based on token-level triggers.

    Similar to TriggeredTagsFormat but uses token IDs instead of string triggers.
    Tags must use ``TokenFormat`` ``begin`` fields, not strings.
    For string-level dispatch, use ``TriggeredTagsFormat`` instead.
    """

    type: Literal["token_triggered_tags"] = "token_triggered_tags"
    """The type of the format."""
    trigger_tokens: List[Union[int, str]]
    """The trigger token IDs or strings."""
    tags: List[TagFormat]
    """The tags to dispatch to."""
    exclude_tokens: List[Union[int, str]] = []
    """List of token IDs or strings to exclude."""
    at_least_one: bool = False
    """Whether at least one tag must be generated."""
    stop_after_first: bool = False
    """Whether to stop after the first tag is generated."""


class TagsWithSeparatorFormat(BaseModel):
    """A format that matches a tags with separator. It can match zero, one, or more tags, separated
    by the separator, with no other text allowed.

    Examples
    --------

    .. code-block:: python

        structural_tag = TagsWithSeparatorFormat(
            tags=[
                TagFormat(begin="<function=func1>", content=JSONSchemaFormat(json_schema=...), end="</function>"),
                TagFormat(begin="<function=func2>", content=JSONSchemaFormat(json_schema=...), end="</function>"),
            ],
            separator=",",
            at_least_one=False,
            stop_after_first=False,
        )

    The above structural tag can accept an empty string, or the following outputs::

        <function=func1>{"name": "John", "age": 30}</function>
        <function=func1>{"name": "John", "age": 30}</function>,<function=func2>{"name": "Jane", "age": 25}</function>
        <function=func1>{"name": "John", "age": 30}</function>,<function=func2>{"name": "Jane", "age": 25}</function>,<function=func1>{"name": "John", "age": 30}</function>
    """

    type: Literal["tags_with_separator"] = "tags_with_separator"
    """The type of the format."""
    tags: List[TagFormat]
    """The tags of the tags with separator."""
    separator: str
    """The separator of the tags with separator."""
    at_least_one: bool = False
    """Whether at least one of the tags must be matched."""
    stop_after_first: bool = False
    """Whether to stop after the first tag is matched."""


class OptionalFormat(BaseModel):
    """A format that matches the content 0 or 1 time (EBNF optional).

    Semantics: the inner format may appear once or not at all.
    """

    type: Literal["optional"] = "optional"
    """The type of the format."""
    content: "Format"
    """The format that may appear 0 or 1 time."""


class PlusFormat(BaseModel):
    """A format that matches the content 1 or more times (EBNF plus).

    Semantics: the inner format must appear at least once.
    """

    type: Literal["plus"] = "plus"
    """The type of the format."""
    content: "Format"
    """The format that must appear at least once."""


class StarFormat(BaseModel):
    """A format that matches the content 0 or more times (EBNF star).

    Semantics: the inner format may appear any number of times.
    """

    type: Literal["star"] = "star"
    """The type of the format."""
    content: "Format"
    """The format that may appear 0 or more times."""


class RepeatFormat(BaseModel):
    """A format that matches the content between min and max times (inclusive).

    Use max=-1 for unbounded upper limit (e.g. "at least min times").
    """

    type: Literal["repeat"] = "repeat"
    """The type of the format."""
    min: int
    """Minimum number of occurrences (inclusive)."""
    max: int
    """Maximum number of occurrences (inclusive). Use -1 for unbounded."""
    content: "Format"
    """The format that is repeated."""


class TagDispatchPair(BaseModel):
    """A (trigger string, content format) pair for TagDispatchFormat."""

    trigger: str
    """The trigger string. When this prefix is seen, generation dispatches to content."""
    content: "Format"
    """The format for the content after the trigger."""


class TagDispatchFormat(BaseModel):
    """A format that maps directly to a TagDispatch grammar.

    Accepts a list of (trigger string, content format) pairs. When the output matches a trigger
    string, generation continues with the corresponding content format. This is a lower-level
    alternative to ``TriggeredTagsFormat`` when you want to specify exact trigger→rule mapping
    without tag begin/end structure.
    """

    type: Literal["tag_dispatch"] = "tag_dispatch"
    """The type of the format."""
    pairs: List[TagDispatchPair]
    """List of (trigger, content) pairs. Trigger must be a non-empty string."""
    loop_after_dispatch: bool = True
    """If true, after handling one dispatch the grammar allows any text until the next trigger."""
    excludes: List[str] = []
    """List of strings that must not appear before a trigger is seen."""


class TokenTagDispatchPair(BaseModel):
    """A (trigger token, content format) pair for TokenTagDispatchFormat."""

    trigger: Union[int, str]
    """The trigger token (ID or token string, resolved via tokenizer)."""
    content: "Format"
    """The format for the content after the trigger token."""


class TokenTagDispatchFormat(BaseModel):
    """A format that maps directly to a TokenTagDispatch grammar.

    Accepts a list of (trigger token, content format) pairs. When the model generates the
    trigger token, generation continues with the corresponding content format. This is a
    lower-level alternative to ``TokenTriggeredTagsFormat`` when you want exact token→rule
    mapping without tag begin/end structure.
    """

    type: Literal["token_tag_dispatch"] = "token_tag_dispatch"
    """The type of the format."""
    pairs: List[TokenTagDispatchPair]
    """List of (trigger token, content) pairs. Trigger can be token ID (int) or token string."""
    loop_after_dispatch: bool = True
    """If true, after one dispatch the grammar allows more tokens until the next trigger."""
    exclude_tokens: List[Union[int, str]] = []
    """Token IDs or strings to exclude before a trigger is seen."""


# ---------- Discriminated Union ----------


Format = Annotated[
    Union[
        AnyTextFormat,
        ConstStringFormat,
        JSONSchemaFormat,
        GrammarFormat,
        RegexFormat,
        QwenXMLParameterFormat,
        OrFormat,
        SequenceFormat,
        TagFormat,
        TriggeredTagsFormat,
        TokenTriggeredTagsFormat,
        TagsWithSeparatorFormat,
        OptionalFormat,
        PlusFormat,
        StarFormat,
        TokenFormat,
        ExcludeTokenFormat,
        AnyTokensFormat,
        RepeatFormat,
        TagDispatchFormat,
        TokenTagDispatchFormat,
    ],
    Field(discriminator="type"),
]
"""Union of all structural tag formats."""


# Solve forward references
if hasattr(BaseModel, "model_rebuild"):
    SequenceFormat.model_rebuild()
    TagFormat.model_rebuild()
    TriggeredTagsFormat.model_rebuild()
    TokenTriggeredTagsFormat.model_rebuild()
    TagsWithSeparatorFormat.model_rebuild()
    OptionalFormat.model_rebuild()
    PlusFormat.model_rebuild()
    StarFormat.model_rebuild()
    RepeatFormat.model_rebuild()
    TagDispatchPair.model_rebuild()
    TagDispatchFormat.model_rebuild()
    TokenTagDispatchPair.model_rebuild()
    TokenTagDispatchFormat.model_rebuild()
elif hasattr(BaseModel, "update_forward_refs"):
    SequenceFormat.update_forward_refs()
    TagFormat.update_forward_refs()
    TriggeredTagsFormat.update_forward_refs()
    TokenTriggeredTagsFormat.update_forward_refs()
    TagsWithSeparatorFormat.update_forward_refs()
    OptionalFormat.update_forward_refs()
    PlusFormat.update_forward_refs()
    StarFormat.update_forward_refs()
    RepeatFormat.update_forward_refs()
    TagDispatchPair.update_forward_refs()
    TagDispatchFormat.update_forward_refs()
    TokenTagDispatchPair.update_forward_refs()
    TokenTagDispatchFormat.update_forward_refs()
else:
    raise RuntimeError("Unsupported pydantic version")


# ---------- Top Level ----------


class StructuralTagItem(BaseModel):
    """Deprecated. Definition of a structural tag item.

    See :meth:`xgrammar.Grammar.from_structural_tag` for more details.
    """

    begin: str
    """The begin tag."""
    schema_: Union[str, Type[BaseModel], Dict[str, Any]] = Field(alias="schema")
    """The schema."""
    end: str
    """The end tag."""


class StructuralTag(BaseModel):
    """
    Describes a complete structural tag structure. It corresponds to
    ``"response_format": {"type": "structural_tag", "format": {...}}`` in API.
    """

    type: Literal["structural_tag"] = "structural_tag"
    """The type must be "structural_tag"."""
    format: Format
    """The format of the structural tag. Could be any of the structural tag formats."""

    @staticmethod
    def from_legacy_structural_tag(
        tags: List[StructuralTagItem], triggers: List[str]
    ) -> "StructuralTag":
        """Convert a legacy structural tag item to a structural tag."""
        return StructuralTag(
            type="structural_tag",
            format=TriggeredTagsFormat(
                type="triggered_tags",
                triggers=triggers,
                tags=[
                    TagFormat(
                        begin=tag.begin,
                        content=JSONSchemaFormat(
                            json_schema=(
                                json.loads(tag.schema_)
                                if isinstance(tag.schema_, str)
                                else (
                                    tag.schema_.model_json_schema()
                                    if isinstance(tag.schema_, type)
                                    and issubclass(tag.schema_, BaseModel)
                                    else tag.schema_
                                )
                            )
                        ),
                        end=tag.end,
                    )
                    for tag in tags
                ],
            ),
        )

    @staticmethod
    def from_json(json_str: Union[str, Dict[str, Any]]) -> "StructuralTag":
        """Convert a JSON string to a structural tag."""
        if isinstance(json_str, str):
            return StructuralTag.model_validate_json(json_str)
        elif isinstance(json_str, dict):
            return StructuralTag.model_validate(json_str)
        else:
            raise ValueError("Invalid JSON string or dictionary")


__all__ = [
    "ConstStringFormat",
    "JSONSchemaFormat",
    "QwenXMLParameterFormat",
    "AnyTextFormat",
    "GrammarFormat",
    "RegexFormat",
    "TokenFormat",
    "ExcludeTokenFormat",
    "AnyTokensFormat",
    "SequenceFormat",
    "OrFormat",
    "TagFormat",
    "TriggeredTagsFormat",
    "TokenTriggeredTagsFormat",
    "TagsWithSeparatorFormat",
    "OptionalFormat",
    "PlusFormat",
    "StarFormat",
    "RepeatFormat",
    "Format",
    "StructuralTagItem",
    "StructuralTag",
]

File: rust/src/structural_tag.rs
//! Structural tag definitions, the Rust counterpart of
//! `xgrammar.structural_tag` (pydantic models in Python, serde here).
//!
//! The only contract with the C++ engine is the serialized JSON produced by
//! [`StructuralTag::to_json`]; the wire format is identical to what the
//! pydantic models emit, so tags serialized by either language are accepted
//! by the other.
//!
//! Python's per-format classes map onto [`Format`] variants:
//! `ConstStringFormat(value="x")` becomes `Format::ConstString { value }`,
//! and so on. [`TagFormat`] stays a standalone struct because several formats
//! embed lists of tags.

use serde::{Deserialize, Deserializer, Serialize, Serializer};

use crate::error::Result;

/// A token reference: either a token id or the token's string form
/// (Python's `Union[int, str]`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum TokenRef {
    /// A token id.
    Id(i64),
    /// The token string.
    Text(String),
}

impl From<i64> for TokenRef {
    fn from(id: i64) -> Self {
        TokenRef::Id(id)
    }
}

impl From<&str> for TokenRef {
    fn from(text: &str) -> Self {
        TokenRef::Text(text.to_string())
    }
}

/// How [`Format::JsonSchema`] content is rendered (JSON or a model-specific
/// XML dialect).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum JsonSchemaStyle {
    /// Standard JSON.
    #[default]
    Json,
    /// Qwen XML: `<parameter=key>value</parameter>`.
    QwenXml,
    /// MiniMax XML: `<parameter name="key">value</parameter>`.
    MinimaxXml,
    /// DeepSeek XML (DeepSeek-v3.2).
    DeepseekXml,
    /// GLM XML: `<arg_key>key</arg_key><arg_value>value</arg_value>`.
    GlmXml,
}

/// The begin marker of a [`TagFormat`]: a string or a single token.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum TagBegin {
    /// A literal begin string.
    Text(String),
    /// A single-token begin marker (`{"type": "token", ...}` on the wire).
    Token(TokenMarker),
}

impl From<&str> for TagBegin {
    fn from(text: &str) -> Self {
        TagBegin::Text(text.to_string())
    }
}

/// The end marker(s) of a [`TagFormat`]: a string, several alternative
/// strings, or a single token.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum TagEnd {
    /// A literal end string.
    Text(String),
    /// Alternative end strings; any of them closes the tag.
    AnyOf(Vec<String>),
    /// A single-token end marker.
    Token(TokenMarker),
}

impl From<&str> for TagEnd {
    fn from(text: &str) -> Self {
        TagEnd::Text(text.to_string())
    }
}

/// A single-token format used inside [`TagBegin`]/[`TagEnd`]
/// (`{"type": "token", "token": ...}` on the wire, Python's `TokenFormat`).
#[derive(Debug, Clone, PartialEq)]
pub struct TokenMarker {
    /// The token id or string.
    pub token: TokenRef,
}

impl Serialize for TokenMarker {
    fn serialize<S: Serializer>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error> {
        #[derive(Serialize)]
        struct Wire<'a> {
            #[serde(rename = "type")]
            type_: &'static str,
            token: &'a TokenRef,
        }
        Wire {
            type_: "token",
            token: &self.token,
        }
        .serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for TokenMarker {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> std::result::Result<Self, D::Error> {
        #[derive(Deserialize)]
        struct Wire {
            #[serde(rename = "type")]
            type_: String,
            token: TokenRef,
        }
        let wire = Wire::deserialize(deserializer)?;
        if wire.type_ != "token" {
            return Err(serde::de::Error::custom(format!(
                "expected type \"token\", got {:?}",
                wire.type_
            )));
        }
        Ok(TokenMarker { token: wire.token })
    }
}

/// A tag: `begin content end` (Python's `TagFormat`).
#[derive(Debug, Clone, PartialEq)]
pub struct TagFormat {
    /// The begin marker.
    pub begin: TagBegin,
    /// The content between the markers.
    pub content: Box<Format>,
    /// The end marker(s).
    pub end: TagEnd,
}

impl TagFormat {
    /// Convenience constructor.
    pub fn new(begin: impl Into<TagBegin>, content: Format, end: impl Into<TagEnd>) -> Self {
        Self {
            begin: begin.into(),
            content: Box::new(content),
            end: end.into(),
        }
    }
}

impl Serialize for TagFormat {
    fn serialize<S: Serializer>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error> {
        #[derive(Serialize)]
        struct Wire<'a> {
            #[serde(rename = "type")]
            type_: &'static str,
            begin: &'a TagBegin,
            content: &'a Format,
            end: &'a TagEnd,
        }
        Wire {
            type_: "tag",
            begin: &self.begin,
            content: &self.content,
            end: &self.end,
        }
        .serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for TagFormat {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> std::result::Result<Self, D::Error> {
        #[derive(Deserialize)]
        struct Wire {
            #[serde(rename = "type", default)]
            type_: Option<String>,
            begin: TagBegin,
            content: Box<Format>,
            end: TagEnd,
        }
        let wire = Wire::deserialize(deserializer)?;
        if let Some(t) = &wire.type_ {
            if t != "tag" {
                return Err(serde::de::Error::custom(format!(
                    "expected type \"tag\", got {t:?}"
                )));
            }
        }
        Ok(TagFormat {
            begin: wire.begin,
            content: wire.content,
            end: wire.end,
        })
    }
}

impl From<TagFormat> for Format {
    fn from(tag: TagFormat) -> Self {
        Format::Tag {
            begin: tag.begin,
            content: tag.content,
            end: tag.end,
        }
    }
}

fn default_true() -> bool {
    true
}

/// A structural tag format, the discriminated union of every format kind
/// (Python's `Format`). The serialized form carries the kind in the `type`
/// field.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Format {
    /// Matches any text, optionally excluding some substrings.
    AnyText {
        /// Strings that must not appear in the matched text.
        #[serde(default)]
        excludes: Vec<String>,
    },
    /// Matches a constant string.
    ConstString {
        /// The constant string.
        value: String,
    },
    /// Matches JSON (or a model-specific XML dialect) following a schema.
    JsonSchema {
        /// The JSON schema: `true`/`false` or a schema object.
        json_schema: serde_json::Value,
        /// The rendering style.
        #[serde(default)]
        style: JsonSchemaStyle,
        /// Whether object properties may appear in any order.
        #[serde(default)]
        any_order: bool,
        /// Max consecutive whitespace characters; `None` means no limit.
        #[serde(default)]
        max_whitespace_cnt: Option<i64>,
    },
    /// Matches an EBNF grammar.
    Grammar {
        /// The EBNF grammar text.
        grammar: String,
    },
    /// Matches a regular expression.
    Regex {
        /// The regex pattern.
        pattern: String,
    },
    /// Deprecated: use [`Format::JsonSchema`] with
    /// [`JsonSchemaStyle::QwenXml`]. Kept so old serialized tags load.
    QwenXmlParameter {
        /// The JSON schema for the function-call parameters.
        json_schema: serde_json::Value,
    },
    /// Matches one of the alternatives.
    Or {
        /// The alternatives.
        elements: Vec<Format>,
    },
    /// Matches a sequence of formats.
    Sequence {
        /// The sequence elements.
        elements: Vec<Format>,
    },
    /// Matches `begin content end`; see [`TagFormat`].
    Tag {
        /// The begin marker.
        begin: TagBegin,
        /// The content between the markers.
        content: Box<Format>,
        /// The end marker(s).
        end: TagEnd,
    },
    /// Free text until a trigger string dispatches into one of the tags.
    TriggeredTags {
        /// The trigger strings; each tag's begin must extend exactly one
        /// trigger.
        triggers: Vec<String>,
        /// The dispatchable tags (string `begin`s only).
        tags: Vec<TagFormat>,
        /// Whether at least one tag must be generated.
        #[serde(default)]
        at_least_one: bool,
        /// Whether matching stops after the first tag.
        #[serde(default)]
        stop_after_first: bool,
        /// Strings that must not appear in the free text.
        #[serde(default)]
        excludes: Vec<String>,
    },
    /// Like `TriggeredTags`, but dispatching on trigger tokens
    /// (token `begin`s only).
    TokenTriggeredTags {
        /// The trigger tokens.
        trigger_tokens: Vec<TokenRef>,
        /// The dispatchable tags.
        tags: Vec<TagFormat>,
        /// Tokens that must not appear in the free part.
        #[serde(default)]
        exclude_tokens: Vec<TokenRef>,
        /// Whether at least one tag must be generated.
        #[serde(default)]
        at_least_one: bool,
        /// Whether matching stops after the first tag.
        #[serde(default)]
        stop_after_first: bool,
    },
    /// Zero or more tags separated by `separator`, with no other text.
    TagsWithSeparator {
        /// The allowed tags.
        tags: Vec<TagFormat>,
        /// The separator between consecutive tags.
        separator: String,
        /// Whether at least one tag must be matched.
        #[serde(default)]
        at_least_one: bool,
        /// Whether matching stops after the first tag.
        #[serde(default)]
        stop_after_first: bool,
    },
    /// The content appears 0 or 1 time.
    Optional {
        /// The optional content.
        content: Box<Format>,
    },
    /// The content appears 1 or more times.
    Plus {
        /// The repeated content.
        content: Box<Format>,
    },
    /// The content appears 0 or more times.
    Star {
        /// The repeated content.
        content: Box<Format>,
    },
    /// The content appears between `min` and `max` times (inclusive).
    Repeat {
        /// Minimum number of occurrences.
        min: i64,
        /// Maximum number of occurrences; `-1` means unbounded.
        max: i64,
        /// The repeated content.
        content: Box<Format>,
    },
    /// Matches a single token.
    Token {
        /// The token id or string.
        token: TokenRef,
    },
    /// Matches a single token outside the excluded set.
    ExcludeToken {
        /// Tokens that must not be matched.
        #[serde(default)]
        exclude_tokens: Vec<TokenRef>,
    },
    /// Matches zero or more tokens outside the excluded set.
    AnyTokens {
        /// Tokens that must not appear.
        #[serde(default)]
        exclude_tokens: Vec<TokenRef>,
    },
    /// Free text where generating a pattern string dispatches into the
    /// corresponding format.
    Dispatch {
        /// `(pattern, content format)` pairs.
        rules: Vec<(String, Format)>,
        /// Whether to keep matching after the first dispatch.
        #[serde(default = "default_true", rename = "loop")]
        loop_: bool,
        /// Strings that must not appear in the free text.
        #[serde(default)]
        excludes: Vec<String>,
    },
    /// Free text where generating a pattern token dispatches into the
    /// corresponding format.
    TokenDispatch {
        /// `(pattern token, content format)` pairs.
        rules: Vec<(TokenRef, Format)>,
        /// Whether to keep matching after the first dispatch.
        #[serde(default = "default_true", rename = "loop")]
        loop_: bool,
        /// Tokens that must not appear in the free text.
        #[serde(default)]
        exclude_tokens: Vec<TokenRef>,
    },
}

/// Deprecated legacy structural tag item, kept for
/// [`StructuralTag::from_legacy`] (Python's `StructuralTagItem`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct StructuralTagItem {
    /// The begin tag.
    pub begin: String,
    /// The JSON schema (an object, or `true`/`false`).
    #[serde(rename = "schema")]
    pub schema: serde_json::Value,
    /// The end tag.
    pub end: String,
}

/// A complete structural tag, corresponding to
/// `"response_format": {"type": "structural_tag", "format": {...}}` in the
/// OpenAI-style API (Python's `StructuralTag`).
#[derive(Debug, Clone, PartialEq)]
pub struct StructuralTag {
    /// The root format.
    pub format: Format,
}

impl StructuralTag {
    /// Wrap a [`Format`] into a structural tag.
    pub fn new(format: Format) -> Self {
        Self { format }
    }

    /// Serialize to the JSON wire format understood by
    /// [`crate::Grammar::from_structural_tag_json`] and the Python API.
    pub fn to_json(&self) -> Result<String> {
        Ok(serde_json::to_string(self)?)
    }

    /// Parse from the JSON wire format (Python's `StructuralTag.from_json`).
    pub fn from_json(json: &str) -> Result<Self> {
        Ok(serde_json::from_str(json)?)
    }

    /// Convert legacy `(tags, triggers)` structural tags into the modern
    /// representation (Python's `StructuralTag.from_legacy_structural_tag`).
    pub fn from_legacy(tags: &[StructuralTagItem], triggers: &[String]) -> Self {
        StructuralTag::new(Format::TriggeredTags {
            triggers: triggers.to_vec(),
            tags: tags
                .iter()
                .map(|tag| {
                    TagFormat::new(
                        tag.begin.as_str(),
                        Format::JsonSchema {
                            json_schema: tag.schema.clone(),
                            style: JsonSchemaStyle::Json,
                            any_order: false,
                            max_whitespace_cnt: None,
                        },
                        tag.end.as_str(),
                    )
                })
                .collect(),
            at_least_one: false,
            stop_after_first: false,
            excludes: Vec::new(),
        })
    }
}

impl Serialize for StructuralTag {
    fn serialize<S: Serializer>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error> {
        #[derive(Serialize)]
        struct Wire<'a> {
            #[serde(rename = "type")]
            type_: &'static str,
            format: &'a Format,
        }
        Wire {
            type_: "structural_tag",
            format: &self.format,
        }
        .serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for StructuralTag {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> std::result::Result<Self, D::Error> {
        #[derive(Deserialize)]
        struct Wire {
            #[serde(rename = "type")]
            type_: String,
            format: Format,
        }
        let wire = Wire::deserialize(deserializer)?;
        if wire.type_ != "structural_tag" {
            return Err(serde::de::Error::custom(format!(
                "expected type \"structural_tag\", got {:?}",
                wire.type_
            )));
        }
        Ok(StructuralTag {
            format: wire.format,
        })
    }
}
